### PR TITLE
diptest 0.8.0 ❄️ 

### PR DESCRIPTION
**Pull request by msentissi**
diptest 0.8.0 :snowflake:

**Destination channel:** Snowflake

### Links

- [PKG-4743](https://anaconda.atlassian.net/browse/PKG-4743) 
- [Upstream repository](https://github.com/RUrlus/diptest/tree/v0.8.0)

### Explanation of changes:

- Created new feedstock using the github release as a source for testing
- Fixed dependencies/versions

[PKG-4743]: https://anaconda.atlassian.net/browse/PKG-4743?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by ryanskeith — COMMENTED** [review]
These look okay to me.

I am a little confused.  When I was looking at the logs I found that the openmp mutex was added.  Then in the build logs:

INFO     -- diptest | OpenMP disabled: OFF                                                                                                                                        
INFO     -- Found OpenMP_CXX: -fopenmp (found version "4.5")                                                                                                                      
INFO     -- Found OpenMP: TRUE (found version "4.5") 

I am confused at how it found a 4.5 version though 5.1 was pinned.

**Comment by skupr-anaconda** [comment]
> These look okay to me.
> 
> I am a little confused. When I was looking at the logs I found that the openmp mutex was added. Then in the build logs:
> 
> INFO -- diptest | OpenMP disabled: OFF INFO -- Found OpenMP_CXX: -fopenmp (found version "4.5") INFO -- Found OpenMP: TRUE (found version "4.5")
> 
> I am confused at how it found a 4.5 version though 5.1 was pinned.

Later In logs on linux-64, I see:
`  INFO (diptest,lib/python3.12/site-packages/diptest/lib/_diptest_core.cpython-312-x86_64-linux-gnu.so): Needed DSO lib/libgomp.so.1 found in defaults/linux-64::_openmp_mutex==5.1=1_gnu`

**Comment by skupr-anaconda** [comment]
Should we use an `openmp` variant on `osx` too? From logs:
```
  -- diptest | OpenMP disabled: OFF
  -- Could NOT find OpenMP_CXX (missing: OpenMP_CXX_FLAGS OpenMP_CXX_LIB_NAMES)
  -- Could NOT find OpenMP (missing: OpenMP_CXX_FOUND)
  -- Could NOT find OpenMP_CXX (missing: OpenMP_CXX_FLAGS OpenMP_CXX_LIB_NAMES)
  -- Could NOT find OpenMP (missing: OpenMP_CXX_FOUND) 
```
Because for `osx` its location is hardcoded here https://github.com/RUrlus/diptest/blob/v0.8.0/mbuild.sh#L12C5-L12C48 and https://github.com/RUrlus/diptest/blob/v0.8.0/src/diptest-core/cmake/FindDependencies.cmake#L7-L10. On my machine `brew --prefix` expands to `/opt/homebrew`. Do we need to patch this?

**Comment by msentissi** [comment]
> These look okay to me.
> 
> I am a little confused. When I was looking at the logs I found that the openmp mutex was added. Then in the build logs:
> 
> INFO -- diptest | OpenMP disabled: OFF INFO -- Found OpenMP_CXX: -fopenmp (found version "4.5") INFO -- Found OpenMP: TRUE (found version "4.5")
> 
> I am confused at how it found a 4.5 version though 5.1 was pinned.

I added _open_mutex=5.1 for after I found that the it was needed and missing at runtime. I am not sure about this, but my understanding was that the _open_mutex version isn't the openmp version.

**Comment by msentissi** [comment]
> Should we use an `openmp` variant on `osx` too? From logs:
> 
> ```
>   -- diptest | OpenMP disabled: OFF
>   -- Could NOT find OpenMP_CXX (missing: OpenMP_CXX_FLAGS OpenMP_CXX_LIB_NAMES)
>   -- Could NOT find OpenMP (missing: OpenMP_CXX_FOUND)
>   -- Could NOT find OpenMP_CXX (missing: OpenMP_CXX_FLAGS OpenMP_CXX_LIB_NAMES)
>   -- Could NOT find OpenMP (missing: OpenMP_CXX_FOUND) 
> ```
> 
> Because for `osx` its location is hardcoded here https://github.com/RUrlus/diptest/blob/v0.8.0/mbuild.sh#L12C5-L12C48 and https://github.com/RUrlus/diptest/blob/v0.8.0/src/diptest-core/cmake/FindDependencies.cmake#L7-L10. On my machine `brew --prefix` expands to `/opt/homebrew`. Do we need to patch this?

On my machine, if I run `brew install libomp`, it is in fact installed under `/opt/homebrew/opt/libomp`. So I think that path is correct, and the build include libomp when I run it locally after running that install command. However, I don't think we can use brew to build our package with libomp, and this path is only attempted in the cmake logic if OpenMP can't be found at first.
I was able to get it to work on `osx-arm64` by adding `llvm-openmp 14.0.6` as a build dependency in [this commit](https://github.com/AnacondaRecipes/diptest-feedstock/pull/1/commits/366ae080a59a7903263a02a37bba900931f4b272). I am now trying to figure out the packages that will get it to work on `osx-64`, where it is still missing.

**Comment by msentissi** [comment]
> > Should we use an `openmp` variant on `osx` too? From logs:
> > ```
> >   -- diptest | OpenMP disabled: OFF
> >   -- Could NOT find OpenMP_CXX (missing: OpenMP_CXX_FLAGS OpenMP_CXX_LIB_NAMES)
> >   -- Could NOT find OpenMP (missing: OpenMP_CXX_FOUND)
> >   -- Could NOT find OpenMP_CXX (missing: OpenMP_CXX_FLAGS OpenMP_CXX_LIB_NAMES)
> >   -- Could NOT find OpenMP (missing: OpenMP_CXX_FOUND) 
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Because for `osx` its location is hardcoded here https://github.com/RUrlus/diptest/blob/v0.8.0/mbuild.sh#L12C5-L12C48 and https://github.com/RUrlus/diptest/blob/v0.8.0/src/diptest-core/cmake/FindDependencies.cmake#L7-L10. On my machine `brew --prefix` expands to `/opt/homebrew`. Do we need to patch this?
> 
> On my machine, if I run `brew install libomp`, it is in fact installed under `/opt/homebrew/opt/libomp`. So I think that path is correct, and the build include libomp when I run it locally after running that install command. However, I don't think we can use brew to build our package with libomp, and this path is only attempted in the cmake logic if OpenMP can't be found at first. I was able to get it to work on `osx-arm64` by adding `llvm-openmp 14.0.6` as a build dependency in [this commit](https://github.com/AnacondaRecipes/diptest-feedstock/pull/1/commits/366ae080a59a7903263a02a37bba900931f4b272). I am now trying to figure out the packages that will get it to work on `osx-64`, where it is still missing.

The most straightforward way was replicating the [pytorch-feedstock](https://github.com/AnacondaRecipes/pytorch-feedstock/blob/master/recipe/meta.yaml) dependency config.

**Review by skupr-anaconda — APPROVED** [review]
LGTM

File: recipe/meta.yaml [hunk 1]
```diff
@@ -1,45 +1,75 @@
...
+  doc_url: https://github.com/RUrlus/diptest#readme
+  license: GPL-3.0
+  license_family: GPL
+  license_file: LICENSE
+  description: A Python/C(++) implementation of Hartigan & Hartigan's dip test for unimodality.
 
 extra:
   recipe-maintainers:
-    - Jrice1317
+    - mohamedlarbisentissi
```
Comment (boldorider4) [inline recipe/meta.yaml:79]: you should actually maintain the original maintainer and at best, if it's a recipe we started (like it's the case here), you may add yourself.

**Comment by boldorider4** [comment]
> Should we use an `openmp` variant on `osx` too? From logs:
> 
> ```
>   -- diptest | OpenMP disabled: OFF
>   -- Could NOT find OpenMP_CXX (missing: OpenMP_CXX_FLAGS OpenMP_CXX_LIB_NAMES)
>   [...]
> ```
> 
> Because for `osx` its location is hardcoded here https://github.com/RUrlus/diptest/blob/v0.8.0/mbuild.sh#L12C5-L12C48 and https://github.com/RUrlus/diptest/blob/v0.8.0/src/diptest-core/cmake/FindDependencies.cmake#L7-L10. On my machine `brew --prefix` expands to `/opt/homebrew`. Do we need to patch this?

I think the location that is searched for OpenMP should still be patched out as long as you will get the error
```
   -- Could NOT find OpenMP_CXX (missing: OpenMP_CXX_FLAGS OpenMP_CXX_LIB_NAMES)
```
The fact that the path is found on your machine when you install with `brew` has nothing to do with where the build system looks for that library, because ultimately you want this build to run on a `conda` environment without having to depend on `brew`.

The path that should be scanned for libraries should be `$PREFIX/lib`, which is the `lib` location within the `host` environment (`/opt/conda/conda-bld/diptest_XXXXXXXXXX/_h_env_placehoder...`)

**Comment by msentissi** [comment]
> hat the path is found on your machine wh

I agree that we need it to run in a `conda` environment without requiring `brew`. 
After including the required OpenMP packages in the recipe, the `-- Could NOT find OpenMP_CXX (missing: OpenMP_CXX_FLAGS OpenMP_CXX_LIB_NAMES)` error disappears, without relying on `brew`. Do you think it is still necessary to patch the cmake logic ?

**Comment by msentissi** [comment]
> > Should we use an `openmp` variant on `osx` too? From logs:
> > ```
> >   -- diptest | OpenMP disabled: OFF
> >   -- Could NOT find OpenMP_CXX (missing: OpenMP_CXX_FLAGS OpenMP_CXX_LIB_NAMES)
> >   [...]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Because for `osx` its location is hardcoded here https://github.com/RUrlus/diptest/blob/v0.8.0/mbuild.sh#L12C5-L12C48 and https://github.com/RUrlus/diptest/blob/v0.8.0/src/diptest-core/cmake/FindDependencies.cmake#L7-L10. On my machine `brew --prefix` expands to `/opt/homebrew`. Do we need to patch this?
> 
> I think the location that is searched for OpenMP should still be patched out as long as you will get the error
> 
> ```
>    -- Could NOT find OpenMP_CXX (missing: OpenMP_CXX_FLAGS OpenMP_CXX_LIB_NAMES)
> ```
> 
> The fact that the path is found on your machine when you install with `brew` has nothing to do with where the build system looks for that library, because ultimately you want this build to run on a `conda` environment without having to depend on `brew`.
> 
> The path that should be scanned for libraries should be `$PREFIX/lib`, which is the `lib` location within the `host` environment (`/opt/conda/conda-bld/diptest_XXXXXXXXXX/_h_env_placehoder...`)

I agree that we need it to run in a `conda` environment without requiring `brew`.
After including the required OpenMP packages in the recipe, the `-- Could NOT find OpenMP_CXX (missing: OpenMP_CXX_FLAGS OpenMP_CXX_LIB_NAMES)` error disappears, without relying on `brew`. Do you think it is still necessary to patch the cmake logic ?

EDIT: After discussing it with @lorepirri, I patched it out in the latest commit to prevent cmake from falling back on a homebrew install of openmp in any case.

File: recipe/meta.yaml (hunk 1 above)
Comment (msentissi) [reply recipe/meta.yaml:79]: The initial name that was in this recipe was part of [the template repository](https://github.com/AnacondaRecipes/template-repository/blob/main/recipe/meta.yaml) that I used to initialize this repo. Should I still maintain it ?

Comment (boldorider4) [reply recipe/meta.yaml:79]: oh, then no. sorry about that.

**Review by boldorider4 — APPROVED** [review]

**Comment by msentissi** [comment]
@cbouss I noticed that we have 2 artifacts per python version for linux-64. This only occurs for this platform. I am not sure what causes this exactly: does it have to do with the way the recipe uses the mkl selectors ? What follows is the relevant section of the post-merge linux-64 logs.

```
                   INFO  INFO :: The inputs making up the hashes for the built  
                         packages are as follows:                               
                   INFO  {                                                      
                   INFO    "diptest-0.8.0-py310h3eb7516_0.tar.bz2": {           
                   INFO      "recipe": {                                        
                   INFO        "blas_impl": "openblas",                         
                   INFO        "channel_targets": "defaults",                   
                   INFO        "cxx_compiler": "gxx",                           
                   INFO        "cxx_compiler_version": "11.2.0",                
                   INFO        "target_platform": "linux-64"                    
                   INFO      }                                                  
                   INFO    },                                                   
                   INFO    "diptest-0.8.0-py310hb23b6be_0.tar.bz2": {           
                   INFO      "recipe": {                                        
                   INFO        "blas_impl": "mkl",                              
                   INFO        "channel_targets": "defaults",                   
                   INFO        "cxx_compiler": "gxx",                           
                   INFO        "cxx_compiler_version": "11.2.0",                
                   INFO        "mkl": "2023.*",                                 
                   INFO        "target_platform": "linux-64"                    
                   INFO      }                                                  
                   INFO    },                                                   
                   INFO    "diptest-0.8.0-py311h3eb7516_0.tar.bz2": {           
                   INFO      "recipe": {                                        
                   INFO        "blas_impl": "openblas",                         
                   INFO        "channel_targets": "defaults",                   
                   INFO        "cxx_compiler": "gxx",                           
                   INFO        "cxx_compiler_version": "11.2.0",                
                   INFO        "target_platform": "linux-64"                    
                   INFO      }                                                  
                   INFO    },                                                   
                   INFO    "diptest-0.8.0-py311hb23b6be_0.tar.bz2": {           
                   INFO      "recipe": {                                        
                   INFO        "blas_impl": "mkl",                              
                   INFO        "channel_targets": "defaults",                   
                   INFO        "cxx_compiler": "gxx",                           
                   INFO        "cxx_compiler_version": "11.2.0",                
                   INFO        "mkl": "2023.*",                                 
                   INFO        "target_platform": "linux-64"                    
                   INFO      }                                                  
                   INFO    },                                                   
                   INFO    "diptest-0.8.0-py312h3eb7516_0.tar.bz2": {           
                   INFO      "recipe": {                                        
                   INFO        "blas_impl": "openblas",                         
                   INFO        "channel_targets": "defaults",                   
                   INFO        "cxx_compiler": "gxx",                           
                   INFO        "cxx_compiler_version": "11.2.0",                
                   INFO        "target_platform": "linux-64"                    
                   INFO      }                                                  
                   INFO    },                                                   
                   INFO    "diptest-0.8.0-py312hb23b6be_0.tar.bz2": {           
                   INFO      "recipe": {                                        
                   INFO        "blas_impl": "mkl",                              
                   INFO        "channel_targets": "defaults",                   
                   INFO        "cxx_compiler": "gxx",                           
                   INFO        "cxx_compiler_version": "11.2.0",                
                   INFO        "mkl": "2023.*",                                 
                   INFO        "target_platform": "linux-64"                    
                   INFO      }                                                  
                   INFO    },                                                   
                   INFO    "diptest-0.8.0-py38h3eb7516_0.tar.bz2": {            
                   INFO      "recipe": {                                        
                   INFO        "blas_impl": "openblas",                         
                   INFO        "channel_targets": "defaults",                   
                   INFO        "cxx_compiler": "gxx",                           
                   INFO        "cxx_compiler_version": "11.2.0",                
                   INFO        "target_platform": "linux-64"                    
                   INFO      }                                                  
                   INFO    },                                                   
                   INFO    "diptest-0.8.0-py38hb23b6be_0.tar.bz2": {            
                   INFO      "recipe": {                                        
                   INFO        "blas_impl": "mkl",                              
                   INFO        "channel_targets": "defaults",                   
                   INFO        "cxx_compiler": "gxx",                           
                   INFO        "cxx_compiler_version": "11.2.0",                
                   INFO        "mkl": "2023.*",                                 
                   INFO        "target_platform": "linux-64"                    
                   INFO      }                                                  
                   INFO    },                                                   
                   INFO    "diptest-0.8.0-py39h3eb7516_0.tar.bz2": {            
                   INFO      "recipe": {                                        
                   INFO        "blas_impl": "openblas",                         
                   INFO        "channel_targets": "defaults",                   
                   INFO        "cxx_compiler": "gxx",                           
                   INFO        "cxx_compiler_version": "11.2.0",                
                   INFO        "target_platform": "linux-64"                    
                   INFO      }                                                  
                   INFO    },                                                   
                   INFO    "diptest-0.8.0-py39hb23b6be_0.tar.bz2": {            
                   INFO      "recipe": {                                        
                   INFO        "blas_impl": "mkl",                              
                   INFO        "channel_targets": "defaults",                   
                   INFO        "cxx_compiler": "gxx",                           
                   INFO        "cxx_compiler_version": "11.2.0",                
                   INFO        "mkl": "2023.*",                                 
                   INFO        "target_platform": "linux-64"                    
                   INFO      }                                                  
                   INFO    }                                                    
                   INFO  }                                                      
                   INFO                                                         
                   INFO                                                         
                   INFO  ######################################################…
```

**Comment by skupr-anaconda** [comment]
> "blas_impl": "openblas",

@msentissi It's because of different `"blas_impl": "openblas"` or `"blas_impl": "mkl"` on `linux-64`

**Comment by lorepirri** [comment]
> > "blas_impl": "openblas",
> 
> @msentissi It's because of different `"blas_impl": "openblas"` or `"blas_impl": "mkl"` on `linux-64`

@skupr-anaconda is it OK then to have two versions (one for `mkl` and one for `openblas`)?

**Comment by skupr-anaconda** [comment]
> > > "blas_impl": "openblas",
> > 
> > 
> > @msentissi It's because of different `"blas_impl": "openblas"` or `"blas_impl": "mkl"` on `linux-64`
> 
> @skupr-anaconda is it OK then to have two versions (one for `mkl` and one for `openblas`)?

Yes, it's fine. You can check, for example, `numexpr 2.8.7` https://anaconda.org/anaconda/numexpr/files and the recipe https://github.com/AnacondaRecipes/numexpr-feedstock/blob/master/recipe/meta.yaml#L23-L24

**Comment by msentissi** [comment]
> > > > "blas_impl": "openblas",
> > > 
> > > 
> > > @msentissi It's because of different `"blas_impl": "openblas"` or `"blas_impl": "mkl"` on `linux-64`
> > 
> > 
> > @skupr-anaconda is it OK then to have two versions (one for `mkl` and one for `openblas`)?
> 
> Yes, it's fine. You can check, for example, `numexpr 2.8.7` https://anaconda.org/anaconda/numexpr/files and the recipe https://github.com/AnacondaRecipes/numexpr-feedstock/blob/master/recipe/meta.yaml#L23-L24

Perfect, thank you for clearing that up @skupr-anaconda !

**Pull request merged**